packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+    promise,
+    resolve,
+    reject,
+  };
+}
+
+type BlockInputState<F extends ForkName> =
+  | {
+      hasBlock: false;
+      hasAllData: false;
+    }
+  | {
+      hasBlock: false;
+      hasAllData: true;
+    }
+  | {
+      hasBlock: true;
+      hasAllData: false;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+    }
+  | {
+      hasBlock: true;
+      hasAllData: true;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+      timeCompleteSec: number;
+    };
+
+abstract class AbstractBlockInput<F extends ForkName = ForkName, TData extends DAData = DAData>
+  implements IBlockInput<F, TData>
+{
+  abstract type: DAType;
+  daOutOfRange: boolean;
+  timeCreated: number;
+
+  forkName: ForkName;
+  slot: Slot;
+  blockRootHex: string;
+  parentRootHex: string;
+
+  abstract state: BlockInputState<F>;
+
+  protected blockPromise = createPromise<SignedBeaconBlock<F>>();
+  protected dataPromise = createPromise<TData>();
+
+  constructor(init: BlockInputInit) {
+    this.daOutOfRange = init.daOutOfRange;
+    this.timeCreated = init.timeCreated;
+    this.forkName = init.forkName;
+    this.slot = init.slot;
+    this.blockRootHex = init.blockRootHex;
+    this.parentRootHex = init.parentRootHex;
+  }
+
+  abstract addBlock(props: AddBlock<F>): void;
+
+  hasBlock(): boolean {
+    return this.state.hasBlock;
+  }
+
+  getBlock(): SignedBeaconBlock<F> {
+    if (!this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlock from BlockInput without a block"
+      );
+    }
+    return this.state.block;
+  }
+
+  getBlockSource(): SourceMeta {
+    if (!this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlockSource from BlockInput without a block"
+      );
+    }
+    return this.state.source;
+  }
+
+  hasAllData(): boolean {
+    return this.state.hasAllData;
+  }
+
+  hasBlockAndAllData(): boolean {
+    return this.state.hasBlock && this.state.hasAllData;
+  }
+
+  getLogMeta(): LogMetaBasic {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
+      slot: this.slot,
+    };
+  }
+
+  getTimeComplete(): number {
+    if (!this.state.hasBlock || !this.state.hasAllData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_TIME_COMPLETE,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getTimeComplete from BlockInput without a block and data"
+      );
+    }
+    return this.state.timeCompleteSec;
+  }
+
+  waitForBlock(timeout: number, signal?: AbortSignal): Promise<SignedBeaconBlock<F>> {
+    if (!this.state.hasBlock) {
+      return withTimeout(() => this.blockPromise.promise, timeout, signal);
+    }
+    return Promise.resolve(this.state.block);
+  }
```

Comment:
I dont think we need to be so fancy here.  Promise.resolve will have same effect as just returning the `blockPromise.promise` if its already resolved

```suggestion
  waitForBlock(timeout: number, signal?: AbortSignal): Promise<SignedBeaconBlock<F>> {
    return withTimeout(() => this.blockPromise.promise, timeout, signal);
  }
```

Reply:
same effect, but if you look at the implementation of withTimeout, it:
- creates an anySignal, which attaches events on both signals
- creates a sleep, which also attaches  events
- wraps the promise in a race
- and triggers the abort controller to detach all events that have been attached

Or we can just check the state and return the value.